# Data Cleaning Challenge

#### Import the base librairies

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


### Import additif librairie

In [2]:
# importation de la bilbiothèque re pour la manipulation des expressions regulières
import re

### load the dataset

In [3]:
df1= pd.read_csv('fifa21_raw_data.csv')
df1.shape

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (76) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


(18979, 77)

In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   photoUrl          18979 non-null  object
 1   LongName          18979 non-null  object
 2   playerUrl         18979 non-null  object
 3   Nationality       18979 non-null  object
 4   Positions         18979 non-null  object
 5   Name              18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Team & Contract   18979 non-null  object
 10  ID                18979 non-null  int64 
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  foot              18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  BP                18979 non-null  object
 16  Growth            18979 non-null  int64 
 17  Joined      

In [5]:
df1.columns

Index(['photoUrl', 'LongName', 'playerUrl', 'Nationality', 'Positions', 'Name',
       'Age', '↓OVA', 'POT', 'Team & Contract', 'ID', 'Height', 'Weight',
       'foot', 'BOV', 'BP', 'Growth', 'Joined', 'Loan Date End', 'Value',
       'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 'PHY', 'Hits

In [6]:
df2 = pd.read_csv('fifa21 raw data v2.csv')
df2.shape

(18979, 77)

In [7]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

In [8]:
# See an overview of the data
df1.sample(5)

photoUrl  \
11666  https://cdn.sofifa.com/players/238/018/21_60.png   
4062   https://cdn.sofifa.com/players/230/480/21_60.png   
4245   https://cdn.sofifa.com/players/212/369/21_60.png   
3131   https://cdn.sofifa.com/players/189/462/21_60.png   
2414   https://cdn.sofifa.com/players/238/274/21_60.png   

                               LongName  \
11666                 Delano van Crooij   
4062    Raphael William Paneira Rochedo   
4245   Eduardo Antônio Machado Teixeira   
3131                     Junior Hoilett   
2414                     Florinel Coman   

                                               playerUrl  Nationality  \
11666  http://sofifa.com/player/238018/delano-van-cro...  Netherlands   
4062   http://sofifa.com/player/230480/raphael-willia...       Brazil   
4245   http://sofifa.com/player/212369/eduardo-antoni...       Brazil   
3131   http://sofifa.com/player/189462/junior-hoilett...       Canada   
2414   http://sofifa.com/player/238274/florinel-coman...      Romania   

      Positions             Name  Age  ↓OVA  POT  \
11666        GK    D. van Crooij   29    64   65   
4062     CAM RM  Raphael Paneira   20    71   71   
4245     CM CAM          Eduardo   27    71   72   
3131      LM RM       J. Hoilett   30    72   72   
2414         LW         F. Coman   22    74   84   

                              Team & Contract  ...     A/W     D/W   IR PAC  \
11666      \n\n\n\nVVV-Venlo\n2015 ~ 2022\n\n  ...  Medium  Medium  1 ★  66   
4062          \n\n\n\nGrêmio\n2019 ~ 2023\n\n  ...  Medium  Medium  1 ★  56   
4245        \n\n\n\nSC Braga\n2018 ~ 2023\n\n  ...  Medium     Low  1 ★  62   
3131    \n\n\n\nCardiff City\n2016 ~ 2021\n\n  ...  Medium  Medium  2 ★  86   
2414   \n\n\n\nFCSB (Steaua)\n2017 ~ 2022\n\n  ...    High     Low  1 ★  90   

       SHO PAS  DRI DEF PHY   Hits  
11666   65  60   63  22  65    \n1  
4062    62  68   75  36  60    \n1  
4245    67  73   74  57  56    \n2  
3131    70  67   74  34  61   \n12  
2414    73  69   75  34  66  \n137  

[5 rows x 77 columns]

In [9]:
df2.iloc[:,:11].head(5)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract  
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  
1   92             \n\n\n\nJuventus  2018 ~ 2022  
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023  
3   91      \n\n\n\nManchester City  2015 ~ 2023  
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022

In [10]:
# First, we copy the df to save an original version
cp_df1 = df1.copy()
cp_df2 = df2.copy()

#We start with the second dataset
df = df2.copy()

In [11]:
df.isnull().sum()

ID              0
Name            0
LongName        0
photoUrl        0
playerUrl       0
             ... 
PAS             0
DRI             0
DEF             0
PHY             0
Hits         2595
Length: 77, dtype: int64

In [12]:
df.duplicated().sum()

0

In [13]:
# drop duplicat columns rows if exist
print(df.shape)
df.drop_duplicates(inplace=True)
print(df.shape)

(18979, 77)
(18979, 77)


In [14]:
# Drop the columns photoUrl, playerUrl, Team & Contract
df.drop(['photoUrl','playerUrl'],axis=1,inplace=True,errors='ignore')

In [15]:
# rename '↓OVA' column to'OVA'
df.rename(columns={'↓OVA':'OVA'},inplace=True,)

In [16]:
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                         Club     Contract   Positions Height  \
0   93   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   
1   92   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   
2   91   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   
3   91   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   
4   91   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   

  Weight Preferred Foot  
0   72kg           Left  
1   83kg          Right  
2   87kg          Right  
3   70kg          Right  
4   68kg          Right

### Clean  'Club' column

In [17]:
# Remove all special caractar in 'Club' Column using regex
df['Club'] = df['Club'].str.replace('\n', '', regex=True)

# Now, we remove whitespace and characters from the beginning and the end of 'Club'
df['Club'] = df['Club'].str.strip()
df.iloc[:,:13].head(5)

ID               Name                      LongName Nationality  Age  \
0  158023           L. Messi                  Lionel Messi   Argentina   33   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro    Portugal   35   
2  200389           J. Oblak                     Jan Oblak    Slovenia   27   
3  192985       K. De Bruyne               Kevin De Bruyne     Belgium   29   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.      Brazil   28   

   OVA  POT                 Club     Contract   Positions Height Weight  \
0   93   93         FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92   92             Juventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   91   93      Atlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91   91      Manchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91   91  Paris Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  
0           Left  
1          Right  
2          Right  
3          Right  
4          Right

### Clean the Contract columns

In [18]:
# We split 'Contract' to 2 collumns, the first is the start year and the second is the End Year
df['Start Year'] = df['Contract'].str.split('~',expand=True)[0]
df['End Year'] = df['Contract'].str.split('~',expand=True)[1]

In [19]:
# Clean the 'Start Year'
df['Start Year'].value_counts()

2020                    5744
2019                    5219
2018                    2749
2017                    1604
2016                     881
Jun 30, 2021 On Loan     770
2015                     524
2014                     371
Free                     237
2013                     205
2012                     135
2011                     102
Dec 31, 2020 On Loan     100
2010                      61
2009                      42
2008                      32
May 31, 2021 On Loan      30
Dec 31, 2021 On Loan      22
2007                      19
Jan 31, 2021 On Loan      18
Jan 1, 2021 On Loan       15
Jun 30, 2022 On Loan      15
2005                      13
2006                      11
2004                       7
Nov 30, 2020 On Loan       7
Jun 23, 2021 On Loan       7
Nov 27, 2021 On Loan       6
2003                       6
Nov 22, 2020 On Loan       5
Jan 3, 2021 On Loan        3
2002                       3
Aug 31, 2021 On Loan       2
Jan 4, 2021 On Loan        2
May 31, 2022 O

**Note 1 :**
>Some start dates are more precise and contain the month and day columns while the majority only contain the year, to keep the information accurate we will reduce all start dates to the year-month-day format; we will do the same for the end date column. In this way, we will be able to calculate more accurately the duration of a player in a team (in year, month or day)

**Note 2 :**
- We have **1013** players in Loan 
- We have **237** players Free

> We will create 2 more columns base on this result : 
- "On Loan" column to indicate if a player is on loan or not
- "Contract Status" column to indicate if a player is In contract or not

In [20]:
df['Start Year'][200:206]

200                   2020 
201                   2020 
202                   2015 
203                   2018 
204                   2007 
205    Jun 30, 2021 On Loan
Name: Start Year, dtype: object

In [21]:
# We create "On Loan" column to indicate if a player is on loan or not

# We set all value off On Loan column to 'No'
df['On Loan'] = 'No'

# We convert 'Start Year' column to string
pd.Series(['Start Year'], dtype="string")

# Now, we check if "On Loan" is in the value of 'Start Year' and set the value of 'On Loan' Column to Yes
for i in range(df.shape[0]):
    if 'On Loan' in df.iloc[i,75]:
        df.iloc[i,77] = 'Yes'

In [22]:
df['On Loan'].value_counts()

No     17966
Yes     1013
Name: On Loan, dtype: int64

In [23]:
# We create "Contract Status" column to indicate if a player is In contract or not

# We set all value off 'Contract Status' column to 'In contract'
df['Contract Status'] = 'In contract'

# We convert 'Start Year' column to string
pd.Series(['Start Year'], dtype="string")

# Now, we check if "Free" is in the value of 'Start Year' and set the value of 'Contract Status' Column to Free
for i in range(df.shape[0]):
    if 'Free' in df.iloc[i,75]:
        df.iloc[i,78] = 'Free'

In [24]:
df['Contract Status'].value_counts()

In contract    18742
Free             237
Name: Contract Status, dtype: int64

In [25]:
# Now we Replace 'On Loan' in the 'Start Year' by whitespace
df['Start Year'] = df['Start Year'].str.replace('On Loan','')

In [26]:
# Now, we remove whitespace and characters from the beginning and the end of 'Start Year'
df['Start Year'] = df['Start Year'].str.strip()

In [27]:
df['Start Year'][200:206]

200            2020
201            2020
202            2015
203            2018
204            2007
205    Jun 30, 2021
Name: Start Year, dtype: object

In [28]:
# Replace month name in column 'Start Year' by the number of the month 
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Jan","01/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Feb","02/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Mar","03/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Apr","04/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("May","05/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Jun","06/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Jul","07/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Aug","08/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Sep","09/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Oct","10/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Nov","11/"))
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace("Dec","12/"))

# Replace ',' in column 'Start Year' by '/'
df['Start Year'] = df['Start Year'].astype('str').map(lambda x : x.replace(",","/"))

# Replace whitespace in column 'Start Year' by ''
df['Start Year'] = df['Start Year'].str.replace(' ', '', regex=True)

In [29]:
df['Start Year'][200:206]

200          2020
201          2020
202          2015
203          2018
204          2007
205    06/30/2021
Name: Start Year, dtype: object

In [30]:
df.iloc[200:210,:]

ID            Name                 LongName         Nationality  Age  \
200  199914           Allan   Allan Marques Loureiro              Brazil   29   
201  181872        A. Vidal             Arturo Vidal               Chile   33   
202  180930        E. Džeko               Edin Džeko  Bosnia Herzegovina   34   
203  178518   R. Nainggolan         Radja Nainggolan             Belgium   32   
204  176676         Marcelo  Marcelo Vieira da Silva              Brazil   32   
205  173731         G. Bale              Gareth Bale               Wales   30   
206  178562       E. Banega              Éver Banega           Argentina   32   
207  172871   J. Vertonghen           Jan Vertonghen             Belgium   33   
208  163705     S. Mandanda           Steve Mandanda              France   35   
209   41236  Z. Ibrahimović       Zlatan Ibrahimović              Sweden   38   

     OVA  POT                    Club              Contract     Positions  \
200   83   83                 Everton           2020 ~ 2023       CM, CDM   
201   83   83                   Inter           2020 ~ 2022  CM, CDM, CAM   
202   83   83                    Roma           2015 ~ 2022            ST   
203   83   83                   Inter           2018 ~ 2022   CAM, CM, CF   
204   83   83             Real Madrid           2007 ~ 2022            LB   
205   83   83       Tottenham Hotspur  Jun 30, 2021 On Loan            RW   
206   83   83               Al Shabab           2020 ~ 2023       CM, CDM   
207   83   83              SL Benfica           2020 ~ 2023        CB, LB   
208   83   83  Olympique de Marseille           2017 ~ 2024            GK   
209   83   83                   Milan           2020 ~ 2021            ST   

     ... SHO PAS DRI  DEF PHY Hits  Start Year End Year On Loan  \
200  ...  71  77  82   85  83  193        2020     2023      No   
201  ...  79  79  78   83  86  128        2020     2022      No   
202  ...  84  71  75   46  73   78        2015     2022      No   
203  ...  80  80  83   83  81  137        2018     2022      No   
204  ...  73  80  88   76  76  133        2007     2022      No   
205  ...  84  82  80   58  73  282  06/30/2021     None     Yes   
206  ...  74  84  82   76  70   61        2020     2023      No   
207  ...  62  72  71   85  78   71        2020     2023      No   
208  ...  81  78  84   48  81   43        2017     2024      No   
209  ...  88  77  79   34  77  425        2020     2021      No   

    Contract Status  
200     In contract  
201     In contract  
202     In contract  
203     In contract  
204     In contract  
205     In contract  
206     In contract  
207     In contract  
208     In contract  
209     In contract  

[10 rows x 79 columns]

**Remark 2:**
> Some "Start Year" values is in format "yyyy", some in format "mm/dd/yyyy" and some others in format "mm/d/yyyy" ; To resolve this, we need to create 3 different regular expression

In [31]:
df['Start Year'].value_counts()

2020          5744
2019          5219
2018          2749
2017          1604
2016           881
06/30/2021     770
2015           524
2014           371
Free           237
2013           205
2012           135
2011           102
12/31/2020     100
2010            61
2009            42
2008            32
05/31/2021      30
12/31/2021      22
2007            19
01/31/2021      18
06/30/2022      15
01/1/2021       15
2005            13
2006            11
06/23/2021       7
11/30/2020       7
2004             7
2003             6
11/27/2021       6
11/22/2020       5
01/3/2021        3
2002             3
05/31/2022       2
08/31/2021       2
01/4/2021        2
07/31/2021       1
12/30/2020       1
1998             1
12/30/2021       1
08/1/2021        1
06/30/2023       1
07/1/2021        1
07/5/2021        1
01/30/2021       1
01/17/2021       1
Name: Start Year, dtype: int64

In [32]:
# We use regular expression to work on "Start Year" to tranform it to "Start date"
year_exp = re.compile(r"([0-9]{4})")
date_exp1 = re.compile(r"([0-9]{2}/[0-9]{2}/[0-9]{4})")
date_exp2 = re.compile(r"([0-9]{2}/[0-9]{1}/[0-9]{4})")

In [33]:
print(date_exp1.fullmatch('06/12/2021'))

<re.Match object; span=(0, 10), match='06/12/2021'>


**Remark 1 :**
>for players who is free, we affected the value '03/13/2040' to "start_date" and "end Date" columns, just to have the same data type in this columns

In [34]:
dt= '01/1/2020'
dt.split('/')

['01', '1', '2020']

In [35]:
# Now, we correct the value of "Start Year" Column
for i in range(df.shape[0]):
    if year_exp.match(df.iloc[i,75]):
        df.iloc[i,75] = '01/01/' + df.iloc[i,75]
    elif date_exp1.match(df.iloc[i,75]):
        df.iloc[i,75] = df.iloc[i,75]
    elif date_exp2.match(df.iloc[i,75]):
        df.iloc[i,75] = df.iloc[i,75].split('/')[0] + '/0' + df.iloc[i,75].split('/')[1]  + df.iloc[i,75].split('/')[2]
    elif df.iloc[i,75] == 'Free':
        # for players who is free, we affected the value '03/13/2040' to "start_date" column, just to have the same data type in this columns
        df.iloc[i,75] = '03/13/2040'

In [36]:
df['Start Year'].value_counts()

01/01/2020    5744
01/01/2019    5219
01/01/2018    2749
01/01/2017    1604
01/01/2016     881
06/30/2021     770
01/01/2015     524
01/01/2014     371
03/13/2040     237
01/01/2013     205
01/01/2012     135
01/01/2011     102
12/31/2020     100
01/01/2010      61
01/01/2009      42
01/01/2008      32
05/31/2021      30
12/31/2021      22
01/01/2007      19
01/31/2021      18
06/30/2022      15
01/012021       15
01/01/2005      13
01/01/2006      11
01/01/2004       7
11/30/2020       7
06/23/2021       7
11/27/2021       6
01/01/2003       6
11/22/2020       5
01/032021        3
01/01/2002       3
05/31/2022       2
01/042021        2
08/31/2021       2
01/17/2021       1
01/01/1998       1
01/30/2021       1
08/012021        1
07/012021        1
06/30/2023       1
12/30/2020       1
12/30/2021       1
07/052021        1
07/31/2021       1
Name: Start Year, dtype: int64

In [37]:
# Some "Start Year" values in in format 'mm/ddyyyy', We reformat them to format 'mm/dd/yyy'
date_exp3 =  re.compile(r"([0-9]{2}/[0-9]{6})")

date_exp3.match('07/052021')

<re.Match object; span=(0, 9), match='07/052021'>

In [38]:
# Now, we correct the value of "Start Year" Column
for i in range(df.shape[0]):
    if date_exp3.match(df.iloc[i,75]):
        df.iloc[i,75] = df.iloc[i,75].split('/')[0] + '/' + df.iloc[i,75].split('/')[1][:2] + '/' + df.iloc[i,75].split('/')[1][-4:]

In [39]:
# Now, we can easily convert it to correct date format "YYYY-MM-DD"
df['Start Year'] = pd.to_datetime(df['Start Year'], format='%m/%d/%Y')

In [40]:
df['Start Year'].value_counts()

2020-01-01    5744
2019-01-01    5219
2018-01-01    2749
2017-01-01    1604
2016-01-01     881
2021-06-30     770
2015-01-01     524
2014-01-01     371
2040-03-13     237
2013-01-01     205
2012-01-01     135
2011-01-01     102
2020-12-31     100
2010-01-01      61
2009-01-01      42
2008-01-01      32
2021-05-31      30
2021-12-31      22
2007-01-01      19
2021-01-31      18
2021-01-01      15
2022-06-30      15
2005-01-01      13
2006-01-01      11
2020-11-30       7
2004-01-01       7
2021-06-23       7
2021-11-27       6
2003-01-01       6
2020-11-22       5
2021-01-03       3
2002-01-01       3
2022-05-31       2
2021-08-31       2
2021-01-04       2
2023-06-30       1
2021-01-17       1
2021-01-30       1
2021-08-01       1
2021-07-05       1
2021-07-31       1
2020-12-30       1
2021-07-01       1
1998-01-01       1
2021-12-30       1
Name: Start Year, dtype: int64

In [42]:
df['End Year'].isna().sum()

1250

In [43]:
df['End Year'].value_counts()

 2021    5959
 2022    4244
 2023    3418
 2024    2206
 2020    1289
 2025     606
 2026       5
 2028       1
 2027       1
Name: End Year, dtype: int64

**Problem :**
> "End Year" column have 1250 null values

In [44]:
df[df['End Year'].isna()].head()

ID            Name                          LongName Nationality  \
205  173731         G. Bale                       Gareth Bale       Wales   
248  193105       A. Areola                   Alphonse Areola      France   
254  200888  Danilo Pereira         Danilo Luís Hélio Pereira    Portugal   
289  230347  Welington Dano    Welington Kauê Dano Nascimento      Brazil   
292  230225  Juiano Mestres  Juan Everton Mestres de Mesquita      Brazil   

     Age  OVA  POT                 Club              Contract Positions  ...  \
205   30   83   83    Tottenham Hotspur  Jun 30, 2021 On Loan        RW  ...   
248   27   82   86               Fulham  Jun 30, 2021 On Loan        GK  ...   
254   28   82   82  Paris Saint-Germain  Jun 30, 2021 On Loan   CDM, CM  ...   
289   20   81   81              No Club                  Free    LB, LM  ...   
292   24   81   81              No Club                  Free   CB, CDM  ...   

    SHO PAS DRI  DEF PHY Hits Start Year End Year On Loan Contract Status  
205  84  82  80   58  73  282 2021-06-30     None     Yes     In contract  
248  79  80  84   55  78   94 2021-06-30     None     Yes     In contract  
254  63  72  73   80  85  285 2021-06-30     None     Yes     In contract  
289  54  76  78   75  74  172 2040-03-13     None      No            Free  
292  61  57  62   82  80   75 2040-03-13     None      No            Free  

[5 rows x 79 columns]

**Remark 3:**
- We have no information on the start and end date of the contract of players on loan. Indeed, when a player is under contract with club A, this club can loan him to another club B. The real starting date of the player's contract is the one with his home club, i.e. club A, and not the starting date of the loan with club B. Similarly, the true end date of the player's contract is the date with club A and not the loan end date.

- As we do not have the 'Start date' and 'End date' information for players on loan, we will assume that these dates are equal to the loan start date and loan end date respectively.

- Especially since we notice that the number of players on loan (1113) + the number of free players (237) = number of players without a contract end date (1250).

- So we don't need the column "Loan Date End" anymore, which we will delete after assigning its values to the column "End Year".

In [45]:
# Now, we correct the value of "End Year" Column
for i in range(df.shape[0]):
    if year_exp.match(str(df.iloc[i,76]).replace(' ','')):
        # If "End Year" is in format yyyy, we convert it to format mm/dd/yyyy , we consider the 31/12 ( the last day of the year)
        df.iloc[i,76] = '01/01/' + str(df.iloc[i,76]).replace(' ','')
    if df.iloc[i,78] == 'Free' :
        #i fplayer is free, Start Year = End Year = '03/13/2040' ( this is a random date, just use to identify this player)
        df.iloc[i,76] = '03/13/2040'
    if df.iloc[i,77] =='Yes' :
        # If player is in Loan, we copy the "Loan Date End" to "End Year"
        df.iloc[i,76] = str(df.iloc[i,16])

In [46]:
# Replace month name in column 'End Year' by the number of the month 
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Jan","01/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Feb","02/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Mar","03/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Apr","04/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("May","05/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Jun","06/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Jul","07/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Aug","08/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Sep","09/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Oct","10/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Nov","11/"))
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace("Dec","12/"))

# Replace ',' in column 'End Year' by '/'
df['End Year'] = df['End Year'].astype('str').map(lambda x : x.replace(",","/"))

# Replace whitespace in column 'End Year' by ''
df['End Year'] = df['End Year'].str.replace(' ', '', regex=True)

In [47]:
# Now, we correct the value of "End Year" Column
for i in range(df.shape[0]):
    if date_exp2.match(df.iloc[i,76]):
        df.iloc[i,76] = df.iloc[i,76].split('/')[0] + '/0' + df.iloc[i,76].split('/')[1]  + df.iloc[i,76].split('/')[2]

In [48]:
df['End Year'].value_counts()

01/01/2021    5959
01/01/2022    4244
01/01/2023    3418
01/01/2024    2206
01/01/2020    1289
06/30/2021     770
01/01/2025     606
03/13/2040     237
12/31/2020     100
05/31/2021      30
12/31/2021      22
01/31/2021      18
01/012021       15
06/30/2022      15
06/23/2021       7
11/30/2020       7
11/27/2021       6
11/22/2020       5
01/01/2026       5
01/032021        3
08/31/2021       2
01/042021        2
05/31/2022       2
01/01/2028       1
01/30/2021       1
12/30/2020       1
01/01/2027       1
07/052021        1
07/31/2021       1
07/012021        1
12/30/2021       1
06/30/2023       1
08/012021        1
01/17/2021       1
Name: End Year, dtype: int64

In [49]:
# Now, we correct the value of "Start Year" Column
for i in range(df.shape[0]):
    if date_exp3.match(df.iloc[i,76]):
        df.iloc[i,76] = df.iloc[i,76].split('/')[0] + '/' + df.iloc[i,76].split('/')[1][:2] + '/' + df.iloc[i,76].split('/')[1][-4:]

In [51]:
# Now, we can easily convert it to correct date format "YYYY-MM-DD"
df['End Year'] = pd.to_datetime(df['End Year'], format='%m/%d/%Y')

In [52]:
df['End Year'].value_counts()

2021-01-01    5974
2022-01-01    4244
2023-01-01    3418
2024-01-01    2206
2020-01-01    1289
2021-06-30     770
2025-01-01     606
2040-03-13     237
2020-12-31     100
2021-05-31      30
2021-12-31      22
2021-01-31      18
2022-06-30      15
2021-06-23       7
2020-11-30       7
2021-11-27       6
2020-11-22       5
2026-01-01       5
2021-01-03       3
2021-01-04       2
2021-08-31       2
2022-05-31       2
2021-07-01       1
2028-01-01       1
2027-01-01       1
2021-08-01       1
2021-12-30       1
2023-06-30       1
2021-07-31       1
2021-07-05       1
2021-01-30       1
2021-01-17       1
2020-12-30       1
Name: End Year, dtype: int64

**Note :**
- It is ok for "Start Year", we can rename it "Start Date"
- It is ok for "End Year", we can rename it "End Date"
- We can drop "Contract" and 'Loan Date End' column

In [246]:
df.rename(columns={"Start Year" :"Start Date", "End Year" :"End Date"},inplace=True)

In [55]:
df.drop(columns=['Loan Date End','Contract'],inplace=True, errors='ignore')

In [57]:
df.iloc[200:210,0:12]

ID            Name                 LongName         Nationality  Age  \
200  199914           Allan   Allan Marques Loureiro              Brazil   29   
201  181872        A. Vidal             Arturo Vidal               Chile   33   
202  180930        E. Džeko               Edin Džeko  Bosnia Herzegovina   34   
203  178518   R. Nainggolan         Radja Nainggolan             Belgium   32   
204  176676         Marcelo  Marcelo Vieira da Silva              Brazil   32   
205  173731         G. Bale              Gareth Bale               Wales   30   
206  178562       E. Banega              Éver Banega           Argentina   32   
207  172871   J. Vertonghen           Jan Vertonghen             Belgium   33   
208  163705     S. Mandanda           Steve Mandanda              France   35   
209   41236  Z. Ibrahimović       Zlatan Ibrahimović              Sweden   38   

     OVA  POT                    Club     Positions Height Weight  \
200   83   83                 Everton       CM, CDM  175cm   74kg   
201   83   83                   Inter  CM, CDM, CAM  180cm   75kg   
202   83   83                    Roma            ST  193cm   80kg   
203   83   83                   Inter   CAM, CM, CF  176cm   65kg   
204   83   83             Real Madrid            LB  174cm   75kg   
205   83   83       Tottenham Hotspur            RW  185cm   82kg   
206   83   83               Al Shabab       CM, CDM  174cm   71kg   
207   83   83              SL Benfica        CB, LB  189cm   86kg   
208   83   83  Olympique de Marseille            GK  185cm   82kg   
209   83   83                   Milan            ST  195cm   95kg   

    Preferred Foot  
200          Right  
201          Right  
202          Right  
203          Right  
204           Left  
205           Left  
206          Right  
207           Left  
208          Right  
209          Right

### Cleaning "Height" Column
- We remove "cm" in value
- We transform "Height" column type to int
- We Rename "Height" to "Height in cm"

In [58]:
df['Height'].isna().sum()

0

In [59]:
df['Height'] = df['Height'].str.replace('cm','')

In [60]:
df['Height'].value_counts()

180     1474
178     1250
185     1182
183     1148
175     1091
        ... 
6'5"       1
5'4"       1
5'6"       1
206        1
155        1
Name: Height, Length: 62, dtype: int64

**Remark:**
- Some values of "Height" column is in feet, and some in cm. We decide to convert all value in cm
> So, Height in cm = feet * 30.48 + inches * 2.54

In [61]:
# We remove the " at the end of some values
df['Height'] = df['Height'].str.replace('"','')
df['Height'].value_counts()[32:]

197     58
198     41
164     31
163     19
199     17
201     13
162     13
6'0      8
160      7
200      7
5'11     6
161      6
6'2      4
5'10     4
202      4
6'3      4
5'9      3
5'7      3
6'1      3
158      3
203      2
159      2
6'4      2
157      1
5'6      1
5'4      1
156      1
155      1
6'5      1
206      1
Name: Height, dtype: int64

In [62]:
feet_re = re.compile(r"([0-9]{1}'[0-9]{1})")
feet_re.match("5'5")

<re.Match object; span=(0, 3), match="5'5">

In [63]:
ch ="5'5"
print(float(ch.split("'")[0]) * 30.48 + float('0.'+ ch.split("'")[1]) * 2.54)
round(float(ch.split("'")[0]) * 30.48 + float('0.' + ch.split("'")[1]) * 2.54)

153.67000000000002


154

# # Some stackoverflow code 
https://stackoverflow.com/questions/67314664/convert-feet-to-cm

A potentially faster way (vectorized) would be using str functions.

The operation you want to perform can be done by applying a dot-product between the 2 elements in your height column and the 2 conversion multipliers.
> [feet, inches] @ [30.48, 2.54] = feet*30.48 + inches*2.54


1. The str.split breaks the string into a list.
2. Then apply(pd.Series) breaks the list into 2 separate columns.
3. Finally, astype(int) converts each of the cells to int
4. Last, dot performs a dot product on each row with the conversion multipliers.

In [ ]:
# conversions = [30.48, 2.54]

# df['new_height'] =''

# for i in range(df.shape[0]):
#     if feet_re.match(df.iloc[i,9]) is None:
#         df.iloc[i,77] = df.iloc[i,9]
#     else:
#         df.iloc[i,77] = round(int(df.iloc[i,9].split("'")[0]) * 30.48 + int(df.iloc[i,9].split("'")[1]) * 2.54)

In [64]:
for i in range(df.shape[0]):
    if feet_re.match(df.iloc[i,9]):
        df.iloc[i,9] = round(int(df.iloc[i,9].split("'")[0]) * 30.48 + int(df.iloc[i,9].split("'")[1]) * 2.54)
    else:
        df.iloc[i,9] = df.iloc[i,9]

In [65]:
df['Height'].value_counts()

180    1474
178    1250
185    1182
183    1148
175    1091
       ... 
157       1
168       1
196       1
156       1
155       1
Name: Height, Length: 62, dtype: int64

In [66]:
df['Height'] = df['Height'].astype(int)

In [67]:
df.rename({'Height':'Height in cm'},inplace=True,axis=1)

**Note :**
> It is ok for the "Height" column

In [68]:
df.iloc[200:210,0:12]

ID            Name                 LongName         Nationality  Age  \
200  199914           Allan   Allan Marques Loureiro              Brazil   29   
201  181872        A. Vidal             Arturo Vidal               Chile   33   
202  180930        E. Džeko               Edin Džeko  Bosnia Herzegovina   34   
203  178518   R. Nainggolan         Radja Nainggolan             Belgium   32   
204  176676         Marcelo  Marcelo Vieira da Silva              Brazil   32   
205  173731         G. Bale              Gareth Bale               Wales   30   
206  178562       E. Banega              Éver Banega           Argentina   32   
207  172871   J. Vertonghen           Jan Vertonghen             Belgium   33   
208  163705     S. Mandanda           Steve Mandanda              France   35   
209   41236  Z. Ibrahimović       Zlatan Ibrahimović              Sweden   38   

     OVA  POT                    Club     Positions  Height in cm Weight  \
200   83   83                 Everton       CM, CDM           175   74kg   
201   83   83                   Inter  CM, CDM, CAM           180   75kg   
202   83   83                    Roma            ST           193   80kg   
203   83   83                   Inter   CAM, CM, CF           176   65kg   
204   83   83             Real Madrid            LB           174   75kg   
205   83   83       Tottenham Hotspur            RW           185   82kg   
206   83   83               Al Shabab       CM, CDM           174   71kg   
207   83   83              SL Benfica        CB, LB           189   86kg   
208   83   83  Olympique de Marseille            GK           185   82kg   
209   83   83                   Milan            ST           195   95kg   

    Preferred Foot  
200          Right  
201          Right  
202          Right  
203          Right  
204           Left  
205           Left  
206          Right  
207           Left  
208          Right  
209          Right

In [69]:
df.Weight.value_counts()

70kg      1495
75kg      1457
80kg      1108
72kg      1022
78kg       991
          ... 
185lbs       1
139lbs       1
190lbs       1
107kg        1
146lbs       1
Name: Weight, Length: 79, dtype: int64

### Cleaning "Weight" Column
- We remove "kg" 
- We delete spaces as a precaution
- We remove lbs in value and transform it to kg
- We Rename "Weight" to "Weight in kg"
> 1 pound (lb) is equal to 0.45359237 kilograms (kg)

In [70]:
df['Weight'] = df['Weight'].str.replace('kg','')
df['Weight'] = df['Weight'].str.replace(' ','')

In [71]:
weight_lb_re = re.compile(r"([0-9]{1,4}lbs)")
weight_lb_re.match("55lbs")

<re.Match object; span=(0, 5), match='55lbs'>

In [72]:
w= '172lbs'

if weight_lb_re.match(w):
    print(round(float(w.replace('lbs','')) * 0.45359237) )

78


In [73]:
for i in range(df.shape[0]):
    if weight_lb_re.match(df.iloc[i,10]):
        df.iloc[i,10] = round(float(df.iloc[i,10].replace('lbs','')) * 0.45359237)
    else:
        df.iloc[i,10] = df.iloc[i,10]

In [74]:
df['Weight'].value_counts()

70     1495
75     1457
80     1108
72     1022
78      991
       ... 
79        1
74        1
110       1
107       1
53        1
Name: Weight, Length: 79, dtype: int64

In [75]:
df['Weight'] = df['Weight'].astype(int)
df.rename({'Weight':'Weight in Kg'},inplace=True,axis=1)

**Note :**
> It is Ok for Wieght column

In [76]:
df['Preferred Foot'].value_counts()

Right    14445
Left      4534
Name: Preferred Foot, dtype: int64

In [77]:
df['BOV'].value_counts()

66    1229
67    1213
65    1193
64    1136
68    1112
69    1019
63     983
70     975
62     859
71     836
72     789
61     768
73     685
60     651
74     565
59     540
75     530
76     398
58     389
57     357
77     354
56     290
78     275
55     263
54     222
53     185
79     185
80     173
52     151
81     134
82      94
51      89
83      73
50      55
84      50
85      43
86      29
87      25
49      22
88      11
48       9
89       8
91       5
90       5
92       1
93       1
Name: BOV, dtype: int64

In [78]:
df['Best Position'].value_counts()

CB     3686
ST     2680
CAM    2299
GK     2075
RM     1611
CDM    1445
LB     1086
RB     1079
CM     1047
LM      871
RW      298
RWB     277
LWB     261
LW      186
CF       78
Name: Best Position, dtype: int64

**Remark :**
- "Joined" in not in the good date format
- We convert it to date format "yyyy-mm-dd"

In [80]:
# Replace month name in column 'Joined' by the number of the month 
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Jan","01/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Feb","02/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Mar","03/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Apr","04/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("May","05/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Jun","06/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Jul","07/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Aug","08/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Sep","09/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Oct","10/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Nov","11/"))
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace("Dec","12/"))

# Replace ',' in column 'Joined' by '/'
df['Joined'] = df['Joined'].astype('str').map(lambda x : x.replace(",","/"))

# Replace whitespace in column 'Joined' by ''
df['Joined'] = df['Joined'].str.replace(' ', '', regex=True)

In [82]:
# Now, we correct the value of "Joined" Column
for i in range(df.shape[0]):
    if date_exp2.match(df.iloc[i,14]):
        df.iloc[i,14] = df.iloc[i,14].split('/')[0] + '/0' + df.iloc[i,14].split('/')[1]  + df.iloc[i,14].split('/')[2]
    if date_exp3.match(df.iloc[i,14]):
        df.iloc[i,14] = df.iloc[i,14].split('/')[0] + '/' + df.iloc[i,14].split('/')[1][:2] + '/' + df.iloc[i,14].split('/')[1][-4:]

In [83]:
# Now, we can easily convert it to correct date format "YYYY-MM-DD"
df['Joined'] = pd.to_datetime(df['Joined'], format='%m/%d/%Y')

In [84]:
df['Joined'].value_counts()

2019-07-01    1344
2018-07-01     865
2019-01-01     682
2017-07-01     520
2020-07-01     435
              ... 
2012-09-04       1
2012-03-30       1
2014-02-04       1
2017-11-01       1
2017-08-27       1
Name: Joined, Length: 1869, dtype: int64

**Note :**
> It is good for "Joined" Column

In [150]:
# df['Value'] = df2['Value']

In [151]:
df['Value'].value_counts()[30:60]

€625K    236
€2.3M    222
€900K    218
€250K    214
€750K    210
€2M      204
€2.2M    184
€675K    180
€775K    180
€925K    168
€800K    167
€975K    165
€875K    163
€2.4M    160
€2.7M    157
€240K    154
€2.5M    153
€5.5M    149
€650K    139
€210K    137
€950K    133
€200K    127
€3M      125
€6.5M    122
€110K    121
€190K    120
€3.1M    117
€2.6M    116
€230K    116
€220K    116
Name: Value, dtype: int64

### Problem in "Value" column
- Some values are in Millions and others in Thousands, 
- We will reduce them to the level of units

### Solution :
- Remove the "€" symbole
- Remove the "M" and convert the value to units
- Remove the "K" and convert the value to units
- Convert column to int

> Convertion Formula
- value in units = value in Millions * 1.000.000
- value in units = value in Thousands * 1.000

In [152]:
df['Value'] = df['Value'].str.replace('€','')
df['Value'] = df['Value'].str.replace(' ','')

In [164]:
df['Value'].value_counts()

1200000.0     582
1100000.0     559
1300000.0     534
1600000.0     532
1000000.0     497
             ... 
63500000.0      1
55000000.0      1
45000000.0      1
95000000.0      1
61500000.0      1
Name: Value, Length: 255, dtype: int64

In [158]:
million_exp = re.compile(r"([0-9]{0,4}.*[0-9]{0,2}M)")

million_exp.match("4M")

<re.Match object; span=(0, 2), match='4M'>

In [159]:
ch = '63M'
float(ch.replace('M','')) * 1000000

63000000.0

In [160]:
thousand_exp = re.compile(r"([0-9]{1,3}K)")

thousand_exp.match("925K")

<re.Match object; span=(0, 4), match='925K'>

In [161]:
ch = "325K"
int(ch.replace('K','')) * 1000

325000

In [162]:
# Now, we correct the value of "Value" Column
for i in range(df.shape[0]):
    if million_exp.match(df.iloc[i,15]):
        df.iloc[i,15] = float( df.iloc[i,15].replace('M','') ) * 1000000
    elif thousand_exp.match(df.iloc[i,15]):
        df.iloc[i,15] = float(df.iloc[i,15].replace('K','')) * 1000
    else:
        df.iloc[i,15] = float(df.iloc[i,15])

In [163]:
df['Value'].value_counts()

1200000.0     582
1100000.0     559
1300000.0     534
1600000.0     532
1000000.0     497
             ... 
63500000.0      1
55000000.0      1
45000000.0      1
95000000.0      1
61500000.0      1
Name: Value, Length: 255, dtype: int64

**Note :**
> We do the same work on "Wage" and "Release Clause" Columns

In [171]:
df['Wage'] = df2['Wage']

In [172]:
df['Wage'].value_counts()

€2K      2899
€500     2042
€1K      1891
€3K      1823
€4K      1050
         ... 
€270K       2
€290K       1
€370K       1
€88K        1
€560K       1
Name: Wage, Length: 134, dtype: int64

In [174]:
df['Wage'] = df['Wage'].str.replace('€','')
df['Wage'] = df['Wage'].str.replace(' ','')

In [175]:
# Now, we correct the value of "Wage" Column
for i in range(df.shape[0]):
    if million_exp.match(df.iloc[i,16]):
        df.iloc[i,16] = float( df.iloc[i,16].replace('M','') ) * 1000000
    elif thousand_exp.match(df.iloc[i,16]):
        df.iloc[i,16] = float(df.iloc[i,16].replace('K','')) * 1000
    else:
        df.iloc[i,16] = float(df.iloc[i,16])

In [176]:
df['Wage'].value_counts()

2000.0      2899
500.0       2042
1000.0      1891
3000.0      1823
4000.0      1050
            ... 
87000.0        2
370000.0       1
560000.0       1
88000.0        1
290000.0       1
Name: Wage, Length: 134, dtype: int64

In [177]:
df['Release Clause'].value_counts()

€0        1261
€1.1M      609
€1.2M      504
€1.3M      443
€1.4M      385
          ... 
€42M         1
€78.9M       1
€54.3M       1
€227K        1
€69.9M       1
Name: Release Clause, Length: 1216, dtype: int64

In [178]:
df['Release Clause'] = df['Release Clause'].str.replace('€','')
df['Release Clause'] = df['Release Clause'].str.replace(' ','')

In [179]:
# Now, we correct the value of "Release Clause" Column
for i in range(df.shape[0]):
    if million_exp.match(df.iloc[i,17]):
        df.iloc[i,17] = float( df.iloc[i,17].replace('M','') ) * 1000000
    elif thousand_exp.match(df.iloc[i,17]):
        df.iloc[i,17] = float(df.iloc[i,17].replace('K','')) * 1000
    else:
        df.iloc[i,17] = float(df.iloc[i,17])

In [180]:
df['Release Clause'].value_counts()

0.0           1261
1100000.0      609
1200000.0      504
1300000.0      443
1400000.0      385
              ... 
27500000.0       1
54200000.0       1
24500000.0       1
81900000.0       1
35000000.0       1
Name: Release Clause, Length: 1216, dtype: int64

**Note :**
> All is ok for "Value", "Wage" and "Release Clause"

In [184]:
df.iloc[:10,12:25]

BOV Best Position     Joined        Value      Wage Release Clause  \
0   93            RW 2004-07-01  103500000.0  560000.0    138400000.0   
1   92            ST 2018-07-10   63000000.0  220000.0     75900000.0   
2   91            GK 2014-07-16  120000000.0  125000.0    159400000.0   
3   91           CAM 2015-08-30  129000000.0  370000.0    161000000.0   
4   91            LW 2017-08-03  132000000.0  270000.0    166500000.0   
5   91            ST 2014-07-01  111000000.0  240000.0    132000000.0   
6   90            RW 2017-07-01  120500000.0  250000.0    144300000.0   
7   90            GK 2018-07-19  102000000.0  160000.0    120300000.0   
8   91            ST 2018-07-01  185500000.0  160000.0    203100000.0   
9   90            GK 2014-07-01  110000000.0  260000.0    147700000.0   

   Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  \
0        429        85         95                70             91       88   
1        437        84         95                90             82       86   
2         95        13         11                15             43       13   
3        407        94         82                55             94       82   
4        408        85         87                62             87       87   
5        423        71         94                85             84       89   
6        392        79         91                59             84       79   
7        114        17         13                19             45       20   
8        408        78         91                73             83       83   
9        118        18         14                11             61       14   

   Skill  
0    470  
1    414  
2    109  
3    441  
4    448  
5    407  
6    406  
7    138  
8    394  
9    144

In [186]:
df.iloc[:10,25:40]

Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
0         96     93           94            91            96       451   
1         88     81           76            77            92       431   
2         12     13           14            40            30       307   
3         88     85           83            93            92       398   
4         95     88           89            81            95       453   
5         85     79           85            70            88       407   
6         90     83           69            75            89       460   
7         27     19           18            44            30       268   
8         92     79           63            70            90       458   
9         21     18           12            63            30       254   

   Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  \
0            91            80       91         94       95    389          86   
1            87            91       87         95       71    444          94   
2            43            60       67         88       49    268          59   
3            77            76       78         91       76    408          91   
4            94            89       96         91       83    357          80   
5            77            78       77         93       82    420          89   
6            94            92       91         92       91    393          80   
7            56            47       40         88       37    240          64   
8            96            96       92         92       82    404          86   
9            38            50       37         86       43    268          66   

   Jumping  Stamina  
0       68       72  
1       95       84  
2       78       41  
3       63       89  
4       62       81  
5       84       76  
6       69       85  
7       52       32  
8       77       86  
9       79       35

In [188]:
df.iloc[:10,40:54]

Strength  Long Shots  Mentality  Aggression  Interceptions  Positioning  \
0        69          94        347          44             40           93   
1        78          93        353          63             29           95   
2        78          12        140          34             19           11   
3        74          91        408          76             66           88   
4        50          84        356          51             36           87   
5        86          85        391          81             49           94   
6        75          84        376          63             55           91   
7        78          14        140          27             11           13   
8        76          79        341          62             38           91   
9        78          10        171          43             22           11   

   Vision  Penalties  Composure  Defending  Marking  Standing Tackle  \
0      95         75         96         91       32               35   
1      82         84         95         84       28               32   
2      65         11         68         57       27               12   
3      94         84         91        186       68               65   
4      90         92         93         94       35               30   
5      79         88         88         96       35               42   
6      84         83         90        122       38               43   
7      66         23         65         50       15               19   
8      80         70         84        100       34               34   
9      70         25         70         48       25               13   

   Sliding Tackle  Goalkeeping  
0              24           54  
1              24           58  
2              18          437  
3              53           56  
4              29           59  
5              19           51  
6              41           62  
7              16          439  
8              32           42  
9              10          439

In [190]:
df.iloc[:10,54:70]

GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  \
0          6           11          15              14            8   
1          7           11          15              14           11   
2         87           92          78              90           90   
3         15           13           5              10           13   
4          9            9          15              15           11   
5         15            6          12               8           10   
6         14           14           9              11           14   
7         86           88          85              91           89   
8         13            5           7              11            6   
9         88           85          88              88           90   

   Total Stats  Base Stats  W/F  SM     A/W     D/W   IR  PAC  SHO  PAS  DRI  
0         2231         466  4 ★  4★  Medium     Low  5 ★   85   92   91   95  
1         2221         464  4 ★  5★    High     Low  5 ★   89   93   81   89  
2         1413         489  3 ★  1★  Medium  Medium  3 ★   87   92   78   90  
3         2304         485  5 ★  4★    High    High  4 ★   76   86   93   88  
4         2175         451  5 ★  5★    High  Medium  5 ★   91   85   86   94  
5         2195         457  4 ★  4★    High  Medium  4 ★   78   91   78   85  
6         2211         470  3 ★  4★    High  Medium  3 ★   93   86   81   90  
7         1389         490  3 ★  1★  Medium  Medium  3 ★   86   88   85   89  
8         2147         466  4 ★  5★    High     Low  3 ★   96   86   78   91  
9         1442         484  4 ★  1★  Medium  Medium  3 ★   88   85   88   90

### Problem in the columns "W/F", "SM" and "IR"
- "★" symbole is in value, we can not calculate avegare, max , min or others claculations with this

### Solution :
- Remove the "★" symbole and whitespace
- Convert column to int

In [204]:
cols = ["W/F","SM","IR"]

for col in cols:
    df[col] = df[col].str.replace("★","").replace(" ",'')
    df[col] = df[col].astype(int)

AttributeError: Can only use .str accessor with string values!

In [193]:
df.iloc[:10,54:70]

GK Diving  GK Handling  GK Kicking  GK Positioning  GK Reflexes  \
0          6           11          15              14            8   
1          7           11          15              14           11   
2         87           92          78              90           90   
3         15           13           5              10           13   
4          9            9          15              15           11   
5         15            6          12               8           10   
6         14           14           9              11           14   
7         86           88          85              91           89   
8         13            5           7              11            6   
9         88           85          88              88           90   

   Total Stats  Base Stats  W/F  SM     A/W     D/W  IR  PAC  SHO  PAS  DRI  
0         2231         466    4   4  Medium     Low   5   85   92   91   95  
1         2221         464    4   5    High     Low   5   89   93   81   89  
2         1413         489    3   1  Medium  Medium   3   87   92   78   90  
3         2304         485    5   4    High    High   4   76   86   93   88  
4         2175         451    5   5    High  Medium   5   91   85   86   94  
5         2195         457    4   4    High  Medium   4   78   91   78   85  
6         2211         470    3   4    High  Medium   3   93   86   81   90  
7         1389         490    3   1  Medium  Medium   3   86   88   85   89  
8         2147         466    4   5    High     Low   3   96   86   78   91  
9         1442         484    4   1  Medium  Medium   3   88   85   88   90

In [195]:
df['A/W'].value_counts()

Medium    12701
High       5288
Low         990
Name: A/W, dtype: int64

In [198]:
df['A/W'] = df['A/W'].astype('category')

In [196]:
df['D/W'].value_counts()

Medium    13956
High       3297
Low        1726
Name: D/W, dtype: int64

In [199]:
df['D/W'] = df['D/W'].astype('category')

In [201]:
df.iloc[:10,70:]

DEF  PHY  Hits Start Year   End Year On Loan Contract Status
0   38   65   771 2004-01-01 2021-01-01      No     In contract
1   35   77   562 2018-01-01 2022-01-01      No     In contract
2   52   90   150 2014-01-01 2023-01-01      No     In contract
3   64   78   207 2015-01-01 2023-01-01      No     In contract
4   36   59   595 2017-01-01 2022-01-01      No     In contract
5   43   82   248 2014-01-01 2023-01-01      No     In contract
6   45   75   246 2017-01-01 2023-01-01      No     In contract
7   51   91   120 2018-01-01 2024-01-01      No     In contract
8   39   76  1.6K 2018-01-01 2022-01-01      No     In contract
9   45   88   130 2014-01-01 2022-01-01      No     In contract

In [220]:
# df['Hits'] = df2['Hits']

In [222]:
df['Hits'].value_counts()[-50:]

1K      1
278     1
478     1
516     1
344     1
952     1
386     1
374     1
538     1
340     1
1.5K    1
348     1
594     1
695     1
279     1
385     1
2.5K    1
421     1
443     1
167     1
343     1
486     1
1.7K    1
566     1
462     1
425     1
706     1
296     1
585     1
399     1
595     1
270     1
331     1
943     1
324     1
363     1
941     1
411     1
541     1
480     1
321     1
654     1
485     1
263     1
267     1
453     1
504     1
216     1
355     1
947     1
Name: Hits, dtype: int64

### Problem in "Hits" column
- Some values are in Thousands and others in units 
- We will convert all them to units

### Solution :
- Remove the "K" symbole and convert the value to units
- Convert column to int

> Convertion Formula
- value in units = value in Thousands * 1.000

In [223]:
hits_thousands = re.compile(r"([0-9]{0,2}.*[0-9]{0,2})")

ch = '5K'
if hits_thousands.match(ch):
    print(float(ch.replace('K','')) * 1000)

5000.0


In [225]:
for i in range(df.shape[0]):
    if hits_thousands.match(str(df.iloc[i,72])):
        df.iloc[i,72] = float(str(df.iloc[i,72]).replace('K','')) * 1000
    else:
        df.iloc[i,72] = float(df.iloc[i,72])

In [239]:
df.iloc[:5,60:]

Base Stats  W/F  SM     A/W     D/W  IR  PAC  SHO  PAS  DRI  DEF  PHY  \
0         466    4   4  Medium     Low   5   85   92   91   95   38   65   
1         464    4   5    High     Low   5   89   93   81   89   35   77   
2         489    3   1  Medium  Medium   3   87   92   78   90   52   90   
3         485    5   4    High    High   4   76   86   93   88   64   78   
4         451    5   5    High  Medium   5   91   85   86   94   36   59   

          Hits Start Year   End Year On Loan Contract Status  
0  771000000.0 2004-01-01 2021-01-01      No     In contract  
1  562000000.0 2018-01-01 2022-01-01      No     In contract  
2  150000000.0 2014-01-01 2023-01-01      No     In contract  
3  207000000.0 2015-01-01 2023-01-01      No     In contract  
4  595000000.0 2017-01-01 2022-01-01      No     In contract

In [234]:
df['Best Position'].value_counts()

CB     3686
ST     2680
CAM    2299
GK     2075
RM     1611
CDM    1445
LB     1086
RB     1079
CM     1047
LM      871
RW      298
RWB     277
LWB     261
LW      186
CF       78
Name: Best Position, dtype: int64

### Small Problem :
- We convert some columns to "category" 
>   categorical_cols = ['Nationality','Positions', 'Club','Best Position','A/W','D/W','On Loan','Contract Status']

In [241]:
categorical_cols = ['Nationality','Positions', 'Club','Best Position','A/W','D/W','On Loan','Contract Status']

for col in categorical_cols:
    df[col] = df[col].astype('category')

In [247]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ID                18979 non-null  int64         
 1   Name              18979 non-null  object        
 2   LongName          18979 non-null  object        
 3   Nationality       18979 non-null  category      
 4   Age               18979 non-null  int64         
 5   OVA               18979 non-null  int64         
 6   POT               18979 non-null  int64         
 7   Club              18979 non-null  category      
 8   Positions         18979 non-null  category      
 9   Height in cm      18979 non-null  int32         
 10  Weight in Kg      18979 non-null  int32         
 11  Preferred Foot    18979 non-null  object        
 12  BOV               18979 non-null  int64         
 13  Best Position     18979 non-null  category      
 14  Joined            1897

In [245]:
df['Value'] = df['Value'].astype(int)
df['Wage'] = df['Wage'].astype(int)
df['Release Clause'] = df['Release Clause'].astype(int)

## Our Dataset is clean
> We save it to CSV file

In [248]:
print('Finals dimensions of our dataset :' ,df.shape)
df.to_csv("fifa21 raw data v2_clean.csv", index=False, encoding='utf-8')
print('Next time, we will see DATA ANALYSIS')

Finals dimensions of our dataset : (18979, 77)
Next time, we will see DATA ANALYSIS
